vllm/v1/core/kv_cache_manager.py

```diff
@@ -368,10 +387,80 @@ def get_num_common_prefix_blocks(
                 break
         return num_common_blocks
 
+    def _free_block_hashes_by_request_id(self, request_id: str) -> None:
+        """Free the block hashes allocated for the request.
+
+        Args:
+            request_id: The request ID to free the block hashes.
+        """
+        self.req_to_block_hashes.pop(request_id, None)
+
     def free_block_hashes(self, request: Request) -> None:
         """Discard the block hashes for the request.
 
         NOTE: Unlike `free`, this method should be called only when the request
         is finished, not when it is preempted.
         """
-        self.req_to_block_hashes.pop(request.request_id, None)
+        self._free_block_hashes_by_request_id(request.request_id)
+
+    def alloc_and_get_external_blocks(
+        self,
+        request: "Request",
+        computed_blocks: list["KVCacheBlock"],
+        num_computed_tokens: int,
+        kv_connector: "KVConnectorBase_V1",
+    ) -> tuple[list["KVCacheBlock"], int]:
+
+        # Check for cache hit.
+        need_to_allocate = kv_connector.get_num_matched_tokens(
+            request, num_computed_tokens)
+        num_allocated_blocks = 0
+
+        # Cache hit: allocate buffer.
+        if need_to_allocate > 0:
+            # HACK: We don't want the scheduler see the blocks are allocated
+            # and associated with the current request. Instead, we want the
+            # scheduler find that the blocks are already allocated and they
+            # are associated with some other requests (i.e., the case of
+            # prefix caching.
+
+            old_req_id = request.request_id
+            request.request_id = f"{old_req_id}-buf-for-kv-connector"
+            allocated_blocks = self.allocate_slots(
+                request,
+                need_to_allocate,
+                computed_blocks,
+                skip_preallocate=True,
+            )
+            self.kv_connector_buffer_req_ids.append(request.request_id)
+            request.request_id = old_req_id
+
+            if allocated_blocks is None:
+                allocated_blocks = []
+
+            # Avoid over-allocating
+            num_expected_blocks = need_to_allocate // self.block_size
+            allocated_blocks = allocated_blocks[:num_expected_blocks]
+
+            # Back-off one block if the external KV is for all tokens
+            if (len(allocated_blocks) + len(computed_blocks)) \
+                    * self.block_size >= len(request.prompt_token_ids):
+                allocated_blocks = allocated_blocks[:-1]
+
+            computed_blocks = computed_blocks + (allocated_blocks or [])
```

Comment:
As this method append the new `allocated_blocks` to the existing prefix cache hit `computed_blocks`, the method name `alloc_and_get_external_blocks` could not be meaningful, maybe `alloc_and_append_external_blocks` would be a little better?